geps/gep-1767/index.md

```diff
@@ -0,0 +1,701 @@
+    //
+    // Output:
+    //   Access-Control-Allow-Origin: *
+    //
+    // For a simple cross-origin interaction, the gateway sets value of the header 
+    // Access-Control-Allow-Origin same as the `Origin` header provided by the client.
+    //
+    // The status code of a successful response to a "preflight" request is an OK status (i.e., 204 or 200). 
+    // For the "preflight" request, if the request `Origin` does not match the configured allowed origins,
+    // the gateway will return a response with error status (e.g., 403). 
+    // Moreover, the gateway will omit the relevant cross-origin response headers.
+    // Then, the client will not send actual cross-origin request.
+    // 
+    // Input:
+    //   Origin: https://foo.example
+    //
+    // Config:
+    //   allowOrigins: ["https://foo.example", "http://foo.example", "https://test.example", "http://test.example"]
+    //
+    // Output:
+    //   Access-Control-Allow-Origin: https://foo.example
+    //
+    // Support: Extended
+    // +listType=set
+    // +kubebuilder:validation:MaxItems=64
+    AllowOrigins []string `json:"allowOrigins,omitempty"`
+ 
+    // AllowCredentials indicates whether the actual cross-origin request 
+    // allows to include credentials.
+    //
+    // The only valid value for the header `Access-Control-Allow-Credentials` is true (case-sensitive).
+    //
+    // The default value of the config AllowCredentials is false.
+    // If the credentials are not allowed in cross-origin requests, 
+    // the gateway will omit the header `Access-Control-Allow-Credentials` entirely.
+    //
+    // Input:
+    //   Origin: https://foo.example
+    //
+    // Config:
+    //   allowCredentials: true
+    //
+    // Output:
+    //   Access-Control-Allow-Origin: https://foo.example
+    //   Access-Control-Allow-Credentials: true
+    //
+    // Support: Extended
+    //
+    // +optional
+    // +kubebuilder:default=false
+    AllowCredentials bool `json:"allowCredentials,omitempty"`
+
+    // AllowMethods indicates which HTTP methods are supported 
+    // for accessing the requested resource.
+    // The method is case-sensitive.
+    //
+    // Config:
+    //   allowMethods: ["GET, PUT, POST, DELETE, PATCH, OPTIONS"]
+    // 
+    // Output:
+    //   Access-Control-Allow-Methods: GET, PUT, POST, DELETE, PATCH, OPTIONS
+    //
+    // A wildcard indicates that the requests with all HTTP methods are allowed.
+    // When responding to a credentialed requests, the gateway must specify 
+    // one or more HTTP methods in the value of the Access-Control-Allow-Methods response header, 
+    // instead of specifying the * wildcard.
+    //
+    // Config:
+    //   allowMethods: ["*"]
+    // 
+    // Output:
+    //   Access-Control-Allow-Methods: ["*"]
+    //
+    // Support: Extended
+    //
+    // +optional
+    // +listType=set
+    // +kubebuilder:validation:MaxItems=16
+    AllowMethods []string `json:"allowMethods,omitempty"`
```

Comment:
This should be an aliased string type, with included constants for all the valid HTTP methods, and should have kubebuilder Enum tags that set the permitted values. That will ensure that users will get immediate feedback if they provide an invalid value.